### ISSUE TITLE
New models in flocker

### ISSUE THREAD

**Issue by mhollanders** [jsocolar/flocker#105]
Hi there,

I'm currently working on occupancy models specifically geared towards passive acoustic monitoring, where sites are allowed to chance occupancy state throughout the acoustic deployment period and the observation process deals with the waiting times. I'm currently working on the manuscript and it would be great if these models were available in existing R packages, for which I think flocker would make the most sense. Are you open to something like this?

Cheers,

Matt

**Comment by abfleishman** [comment]
@mhollanders I too am very interested in models like you describe. Giant +1!

**Comment by jsocolar** [comment]
Let's try to do this, though I cannot make any promises about timescales for getting things done from my end.  If somebody is willing to do a big chunk of the lifting, I can comment to doing any necessary reviews and polishing in a timely manner.

To get started, check out the instructions here https://github.com/jsocolar/flocker/blob/main/inst/rmd/adding_a_family.Rmd

The main thing is to get a sketch of what the likelihood should look like, and then a sketch of what the `flckerdata` format for this model type needs to look like (if different from the format for one of the existing model types).  I recommend against starting more substantial work on the code prior to agreeing on the data format.  Usually, the biggest challenges involve getting the indexing right during the data packing (in R code) and the data unpacking (in Stan code), and it's possible to waste a lot of time here by racing ahead before we agree to the data format.

**Comment by mhollanders** [comment]
Hi @jsocolar, sounds good. For the single season model, the key data structures are (1) the number of detections per site, primary occasion, and secondary occasion and (2) the waiting times between detections. For the multi-season version, there's an additional index of season. For the single season model, the idea is that there is an "overall occupancy" of each site (analogous to the single season occupancy probability) but that occupancy state is allowed to change within a deployment period. So essentially, this is just a zero-inflated (with the extra occupancy probability) dynamic occupancy probability where the observed data are waiting times. I should probably take a look at the current data format looks like for dynamic occupancy, as it should be analogous. For the multiseason version, I doubt that there is already a data type that has all the indices.

**Comment by jsocolar** [comment]
Could you write out the likelihood in math notation?

**Comment by mhollanders** [comment]
Yes. The idea is that there's an overall occupancy state of site $i \in 1:I$ in season $k \in 1:K$; I'll just write it out for $K=1$ for now. Then, within each season, the deployment period of the acoustic devices is split into occasions $j \in 1:J$, where the interval length $\delta$ should be informed by the species ecology. So the overall occupancy state $m_i$ is a Bernoulli trial which is analogous to the single season occupancy probability:

$$
m_i \sim \text{Bernoulli} (\psi_i)
$$

where $\psi_i$ can be modeled flexibly, i.e. with covariates or (spatial) random effects. Then conditional on $m_i = 1$, the within-occasion occupancy state $z_{ij}$ is allowed to change, and is also modeled with a Bernoulli trial (I opted for this over a dynamic occupancy model with colonisation/emigration probabilities):

$$
z_{ij} | m_i \sim \text{Bernoulli} (\chi_{ij}))
$$

where again, $\chi_{ij}$ can vary by site and occasion.

Conditional on $z_{ij} = 1$, the observed waiting times between detections $\bf{y}_{ij}$ are modeled with an exponential distribution:

$$
y_{ij} | z_{ij} \sim \text{Exponential} (\mu_{ij})
$$

Where the last waiting time (which is the total occasion length if no detections were made) is modeled with a complement of the CDF of the exponential distribution. The call rate $\mu_{ij}$ can also be modeled flexibly with a Hawkes process, which addresses the non-independence of calls, i.e. due to animals vocalising repeatedly or to each other.

So basically, instead of assuming the site is occupied throughout the deployment period, the occupancy process is modeled more flexily and all of the detection data is used in the observation process.

**Comment by mhollanders** [comment]
Hi Jacob, 

I'm having a crack at steps 1--3 as outlined above. Pardon me that I'm a bit slow to understand all of this, as I've never worked on an R package before. I have a few questions.

- `make_flocker_data_augmented()` additionally expects at least two sampling events and that the possible values are `1`, `0`, or `NA`. However, the described model can have one sampling event, which is the waiting time from the start to the end of the occasion when no detections were made. Additionally, the possible values are actually the waiting times, not `1`, `0`, or `NA`. 
- I don't know what you mean by "refactoring" the original function. Can you point me to a resource to understand this?

Sorry about this, I'm a bit out of my depth.

Matt

**Comment by jsocolar** [comment]
No worries! Thanks for taking a look.  An initial word of opportunity and caution: I'm really happy to walk you through the development of these functions (like honestly it would be wonderful!), but this is probably only a good idea if you are substantially interested in learning about writing the software.  If your main interest is just in being able to fit the model, then I suspect that this could turn into a painful slog that might peter out rather than ending.

With that said:
- We are splitting the work into two parts.  The first part ONLY creates the two-level occupancy framework, which enables us to model the occupancy of occasions nested inside of sites, but continues to use binary detection/nondetection data over repeat visits (within occasions) for the likelihood at the occasion level.  We will swap the likelihood over to use waiting times in a separate step.  There are two reasons for splitting the work like this.  
    - Trying to do it all at once would be error-prone and would lead to pull requests that are too large to review effectively.  
    - These variations should be useful on their own, and not only in concert.  When doing this work, I don't want to end up with a two-level waiting-times model and nothing else.  I want to end up with two-level models using both ordinary and waiting-times likelihoods, and with waiting-times models both for the one-level and two-level cases.
- "Refactor" in this context means to re-organize the code into useful component functions that are concise, efficient, and reusable.  In particular, if we just wrote `make_flocker_data_twolevel` by borrowing a bunch of code from `make_flocker_data_augmented`, then we'd end up with a whole bunch of code duplication.  In the future, if we spot any bugs or want to modify the code to be more efficient or more general, we would have to do it in two places.  If we don't remember to do that, then we would end up with two functions that do the same thing but in different ways, and it would get confusing to verify correctness.  Here, the thing to do is to put most of the work-horse code into a single function, and then let the other function call that work-horse function internally.  Because the `make_flocker_data_twolevel` is more general than `make_flocker_data_augmented`, my suggestion is to move the main code into `make_flocker_data_twolevel`, and then re-write `make_flocker_data_augmented` to call `make_flocker_data_twolevel` internally, while performing the necessary additional checks for non-raggedness (every species represented on every visit that exists) and appending the augmented zero-data before formatting using `make_flocker_data_twolevel`

**Comment by jsocolar** [comment]
Hi @mhollanders , I've pointed codex at creating the two-level single-season model, and will slowly review the resulting PR over the coming days (or longer).  https://github.com/jsocolar/flocker/pull/117

If all goes well, it seems like it might be feasible now to extend this to two-level multi-season models, and eventually to waiting-time likelihoods.  Is this still of special interest to you?  (If not, I might start with n-mixtures first).
